### PR TITLE
Add support of multiple bridges in bridge token

### PR DESCRIPTION
**Pull request by k1rill-fedoseev**
Closes https://github.com/poanetwork/tokenbridge-contracts/issues/388

### CODE REVIEW

File: contracts/ERC677MultiBridgeToken.sol [hunk 1]
```diff
@@ -0,0 +1,108 @@
+pragma solidity 0.4.24;
+
+import "./ERC677BridgeToken.sol";
+
+/**
+ * @title ERC677MultiBridgeToken
+ * @dev This contract extends ERC677BridgeToken to support several bridge simulteniously
+ */
+contract ERC677MultiBridgeToken is ERC677BridgeToken {
```
Comment (akolotov) [inline contracts/ERC677MultiBridgeToken.sol]: Consider to inherit this new contract by `ERC677BridgeTokenRewardable`. Please elaborate which components could be also affected after such kind of change.

File: contracts/ERC677MultiBridgeToken.sol [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    /**
+    * @dev Overrides super.callAfterTransfer, uses custom checking if an address is a known bridge contract
+    * @param _from tokens sender address
+    * @param _to tokens receiver address
+    * @param _value amount of transferred tokens
+    */
+    function callAfterTransfer(address _from, address _to, uint256 _value) internal {
+        if (AddressUtils.isContract(_to) && !contractFallback(_from, _to, _value, new bytes(0))) {
+            require(!isBridge(_to));
```
Comment (akolotov) [inline contracts/ERC677MultiBridgeToken.sol]: So, the difference is only this line. Does it make sense to implement something like this:

_ERC677BridgeToken_
```solidity
function callAfterTransfer(address _from, address _to, uint256 _value) internal {
    if (AddressUtils.isContract(_to) && !contractFallback(_from, _to, _value, new bytes(0))) {
        require(!isBridge(_to));
        emit ContractFallbackCallFailed(_from, _to, _value);
    }
}

function isBridge(address _address) public view returns (bool) {
    return _address == bridgeContract;
}
```
In this case in _ERC677MultiBridgeToken_ you will need to define only `isBridge` without necessity to duplicate `callAfterTransfer`.

What do you think?

File: contracts/ERC677MultiBridgeToken.sol [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    constructor(string _name, string _symbol, uint8 _decimals) public ERC677BridgeToken(_name, _symbol, _decimals) {
+        bridgePointers[F_ADDR] = F_ADDR; // empty bridge contracts list
+    }
+
+    /**
+    * @dev Adds one more bridge contract into the list
+    * @param _bridge bridge contract address
+    */
+    function setBridge(address _bridge) external onlyOwner {
```
Comment (varasev) [inline contracts/ERC677MultiBridgeToken.sol]: Maybe name this method as `addBridge`?

File: contracts/ERC677MultiBridgeToken.sol [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        bridgeCount = bridgeCount.add(1);
+
+        emit BridgeAdded(_bridge);
+    }
+
+    /**
+    * @dev Removes one existing bridge contract from the list
+    * @param _bridge bridge contract address
+    */
+    function unsetBridge(address _bridge) external onlyOwner {
```
Comment (varasev) [inline contracts/ERC677MultiBridgeToken.sol]: As it was proposed above for renaming `setBridge` to `addBridge`, maybe rename `unsetBridge` to `removeBridge`?

**Comment by varasev** [comment]
@k1rill-fedoseev Thank you for the rapid changes! I left a couple of minor comments.

**Comment by varasev** [comment]
Also, should we override `setBridgeContract` in `ERC677MultiBridgeToken` as follows?

```solidity
function setBridgeContract(address) external {
    revert();
}
```

Because `ERC677BridgeToken.setBridgeContract()` function and `ERC677BridgeToken.bridgeContract()` public getter make no sense in `ERC677MultiBridgeToken`.

**Comment by varasev** [comment]
And maybe set overriden `ERC677MultiBridgeToken.bridgeContract` public getter so that it would return the first bridge from the list:

```solidity
function bridgeContract() public view return (address) {
    address firstBridge = bridgePointers[F_ADDR];
    return (firstBridge == F_ADDR) ? address(0) : firstBridge;
}
```

**Comment by akolotov** [comment]
> And maybe set overriden `ERC677MultiBridgeToken.bridgeContract` public getter so that it would return the first bridge from the list:

There is no obvious use case for such method. So, I would recommend to overwrite it with `revert()`.

File: contracts/ERC677MultiBridgeToken.sol (hunk 3 above)
Comment (k1rill-fedoseev) [reply contracts/ERC677MultiBridgeToken.sol]: Sure, I renamed it.

**Pull request merged**